src/main/java/io/strimzi/kproxy/filter/KrpcRequestFilter.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.strimzi.kproxy.filter;
+
+import org.apache.kafka.common.message.MetadataRequestData;
+import org.apache.kafka.common.message.ProduceRequestData;
+import org.apache.kafka.common.protocol.ApiKeys;
+
+import io.strimzi.kproxy.codec.DecodedRequestFrame;
+
+/**
+ * <p>Interface for {@code *RequestFilter}s.
+ * This interface can be implemented in two ways:
+ * <ul>
+ *     <li>filter classes can (multiply) implement one of the RPC-specific subinterfaces such as {@link ProduceRequestFilter} for a type-safe API</li>
+ *     <li>filter classes can extend {@link KrpcGenericRequestFilter}</li>
+ * </ul>
+ *
+ * <p>When implementing one or more of the {@code *RequestFilter} subinterfaces you need only implement
+ * the {@code on*Request} method(s), unless your filter can avoid deserialization in which case
+ * you can override {@link #shouldDeserializeRequest(ApiKeys, short)} as well.</p>
+ *
+ * <p>When extending {@link KrpcGenericRequestFilter} you need to override {@link #apply(DecodedRequestFrame, KrpcFilterContext)},
+ * and may override {@link #shouldDeserializeRequest(ApiKeys, short)} as well.</p>
+ *
+ * <h3>Guarantees</h3>
+ * <p>Implementors of this API may assume the following:</p>
+ * <ol>
+ *     <li>That each instance of the filter is associated with a single channel</li>
+ *     <li>That {@link #shouldDeserializeRequest(ApiKeys, short)} and
+ *     {@link #apply(DecodedRequestFrame, KrpcFilterContext)} (or {@code on*Request} as appropriate)
+ *     will always be invoked on the same thread.</li>
+ *     <li>That filters are applied in the order they were configured.</li>
+ * </ol>
+ * <p>From 1. and 2. it follows that you can use member variables in your filter to
+ * store channel-local state.</p>
+ *
+ * <p>Implementors should <strong>not</strong> assume:</p>
+ * <ol>
+ *     <li>That filters in the same chain execute on the same thread. Thus inter-filter communication/state
+ *     transfer needs to be thread-safe</li>
+ * </ol>
+ */
+public /* sealed */ interface KrpcRequestFilter extends KrpcFilter /* TODO permits ... */ {
```

Comment:
Should we open an issue for using sealed types once we've moved to Java 17? I know we're not really expecting it to happen very soon, since it seems netty is blocked on some missing _performant_ replacements for `Unsafe`.

Reply:
Sure, can you do it?